File: src/main/java/Day1.kt
import org.junit.Assert

class Day1: Day<IntArray> {
    override fun parse(text: String): IntArray {
        return text.toCharArray().map { Character.digit(it, 10) }.toIntArray()
    }

    override fun part1(input: IntArray): Any {
        return perform(input, { arr, idx -> arr[(idx + 1) % arr.size] })
    }

    override fun part2(input: IntArray): Any {
        return perform(input, { arr, idx -> arr[(idx + arr.size / 2) % arr.size] })
    }

    fun perform(numbers: IntArray, otherOperation: (IntArray, Int) -> Int): Int {
        return numbers.foldIndexed(0, { index, current: Int, value ->
            val other: Int = otherOperation.invoke(numbers, index)
            val add: Int = if (other == value) value else 0
            current + add
        })
    }

}

fun main(args: Array<String>) {
    val d = Day1()

    val text = d.javaClass.classLoader.getResource("day1").readText().split('\n')
    text.filter { it.contains('=') }.forEach {
        val before = it.substring(0, it.indexOf('='))
        val after = it.substring(it.indexOf('=') + 1).trim()
        Assert.assertEquals("Mismatch for input " + before, after.toInt(), d.part1(d.parse(before)))
    }
    val parsed = d.parse(text[text.size - 1])
    println(d.part1(parsed))
    println(d.part2(parsed))
}

File: 2019-typescript/src/Day1.ts
import Day from "./Day";

function fuel(mass: number): number {
  let f = Math.floor(mass / 3) - 2;
  return f > 0 ? f : 0;
}

class Day1 implements Day<number[]> {
  constructor() {}
  parse(input: string): number[] {
    return input.split('\n').map((s: string) => parseInt(s));
  }

  part1(input: number[]): any {
    return input.map(fuel).reduce((a: number, b: number) => a + b, 0);
  }

  part2(input: number[]): any {
    let total = 0;

    input.forEach((m: number) => {
      let f = m;
      do {
        f = fuel(f);
        total += f;
      } while (f > 0);
    });
    return total;
  }
}

export default Day1;
